### PR TITLE
test: add tests for region migration procedure

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

1. Refactors the `State` of the RegionMigration.
2. Add some basic region migration tests. 

And some remaining tests will be added in the next PR.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#2700

### CODE REVIEW

File: src/meta-srv/src/procedure/region_migration/downgrade_leader_region.rs [hunk 1]
```diff
@@ -53,18 +54,25 @@ impl Default for DowngradeLeaderRegion {
...
 #[typetag::serde]
 impl State for DowngradeLeaderRegion {
-    async fn next(&mut self, ctx: &mut Context) -> Result<Box<dyn State>> {
+    async fn next(&mut self, ctx: &mut Context) -> Result<(Box<dyn State>, Status)> {
         // Ensures the `leader_region_lease_deadline` must exist after recovering.
         ctx.volatile_ctx
             .set_leader_region_lease_deadline(Duration::from_secs(REGION_LEASE_SECS));
         self.downgrade_region_with_retry(ctx).await;
 
         // Safety: must exist.
```
Comment (fengjiachun) [inline src/meta-srv/src/procedure/region_migration/downgrade_leader_region.rs:63]: Is this comment need to be removed?

File: src/meta-srv/src/procedure/region_migration/update_metadata/upgrade_candidate_region.rs [hunk 2]
```diff
@@ -76,6 +76,40 @@ impl UpdateMetadata {
...
+                violated: format!("The leader peer of region {region_id} is not found during the update metadata for upgrading"),
+            })?;
+
+        let candidate_peer_id = ctx.persistent_ctx.to_peer.id;
+
+        if leader_peer.id == candidate_peer_id {
+            ensure!(
+                !region_route.is_leader_downgraded(),
+                error::UnexpectedSnafu {
+                    violated: "Unexpected intermediate state is found during the update metadata for upgrading",
```
Comment (fengjiachun) [inline src/meta-srv/src/procedure/region_migration/update_metadata/upgrade_candidate_region.rs:103]: Add `region id`?

**Comment by WenyXu** [comment]
@fengjiachun @MichaelScofield PTAL

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**